Rustdoc - display `since` version for stable items

wesleywiser opened the pull request:
Here's some screenshots after this change:

![screen shot 2016-01-03 at 11 38 30 am](https://cloud.githubusercontent.com/assets/831192/12079661/23da4e38-b20f-11e5-8c84-ba51d7a59c3f.png)
![screen shot 2016-01-03 at 11 40 39 am](https://cloud.githubusercontent.com/assets/831192/12079663/23e00012-b20f-11e5-9f01-408cc8d43687.png)
![screen shot 2016-01-03 at 11 42 17 am](https://cloud.githubusercontent.com/assets/831192/12079662/23dfe6c2-b20f-11e5-9998-53abc643e2ef.png)

I tried to click through the `std` docs and make sure everything that can have stability attributes has it rendered but I'm probably missing some. I'd also appreciate any feedback on the css changes. I had difficulty getting the `since` labels aligning correctly for enum variants. If anyone has a better idea for that, I'd be glad to implement it.

Fixes #27607

sfackler commented:
Yay!

tshepang commented:
I would prefer if the text was less prominent: either grayed or with smaller font.

wesleywiser commented:
@tshepang What do you think of this?
![screen shot 2016-01-03 at 5 14 51 pm](https://cloud.githubusercontent.com/assets/831192/12081089/a2a573c2-b23d-11e5-869d-7bf2742b330e.png)

tshepang commented:
@wesleywiser that's better, thanks

brson commented:
I also have the sense that this is giving a great deal of prominence to a rather unimportant piece of information.

One thing we could do is remove 'since' - this word doesn't convey any information every time it's repeated.

This is also a *new* way to style stability information, and I already count two - See for [example](http://doc.rust-lang.org/std/collections/hash_state/) where stability of the element on the page is indicated by a colored box while stability of its methods is indicated by the word '[Unstable]' while fading out the text. It would be easier to understand that *this* and *that* are both types of stability information if they were styled similarly.

wesleywiser commented:
@brson re 'since': I agree. Here's what it looks like without 'since':
![screen shot 2016-01-05 at 9 42 36 pm](https://cloud.githubusercontent.com/assets/831192/12133668/6a931e44-b3f5-11e5-9436-5990642bef35.png)

I think agree that adding yet another way to show stability information is bad. I'm having some trouble imagining what a common style might look like. 

One idea I had that might further minimize the visual noise would be to only show the version number when it is different than the containing object's version number. For example, the previous screenshot would look like this because `Option` is marked `1.0.0`:
![screen shot 2016-01-05 at 9 48 41 pm](https://cloud.githubusercontent.com/assets/831192/12133749/2f090720-b3f6-11e5-9a13-58df03877883.png)

tshepang commented:
I like the second option:

> ...only show the version number when it is different than the containing object's version number.

brson commented:
@wesleywiser I do like that idea of not displaying them when they are the same. With that I think I'm ok with trying this.

steveklabnik commented:
I basically feel the same as @brson on this thread. Let's give it a try. Let me know when this PR implements that, and we'll merge after the release thursday, assuming you can update it before then.

wesleywiser commented:
@brson @steveklabnik Will do. Sorry, I've been pretty busy lately but hopefully I'll have some time to work on it later this week. I'll post new screenshots when when I update the PR.

steveklabnik commented:
No worries at all :)

steveklabnik commented:
@wesleywiser ping! Any chance you've had the time to work on this?

bluss commented on src/librustdoc/clean/mod.rs:
```diff
@@ -339,6 +339,14 @@ impl Item {
         }
     }
+
+    pub fn stable_since(&self) -> Option<String> {
```
Can you use `-> Option<&str>` here instead? It should be enough since the string is only ever passed down. Change `s.since.clone()` to `&s.since[..]` too and update the other impacted code.

I think this change looks great otherwise.

wesleywiser commented:
@steveklabnik Great timing Steve! I think it's mostly working at this point (I looked over the generated docs and didn't see any glaring issues). I've updated this PR with the new code (and also implemented @bluss's suggestion).

steveklabnik commented:
@bors: r+

okay, let's give it a try! :confetti_ball:

bors commented:
:pushpin: Commit 75acee2 has been approved by `steveklabnik`

<!-- @bors r=steveklabnik 75acee2bdee0278e0dc9610bfd05eb4367e41c20 -->

bors commented:
:hourglass: Testing commit 75acee2 with merge b5da60d...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/8019), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/8040), [auto-linux-64-debug-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/1287), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/7980), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/7990), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/7980), [auto-linux-cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/1193), [auto-linux-musl-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/3145), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/8025), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/8028), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/7993), [auto-mac-ios-opt](http://buildbot.rust-lang.org/builders/auto-mac-ios-opt/builds/163), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/3033), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/3049), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/3040), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/3022), [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/2255), [auto-win-msvc-64-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/3027)
